inode/inode.go

```diff
@@ -351,13 +352,78 @@ func (vS *volumeStruct) inodeCacheTouch(inode *inMemoryInodeStruct) {
 	vS.Unlock()
 }
 
+// TODO: What happens if the volume is removed?  What if
+// failover?
+func (vS *volumeStruct) inodeCacheDiscard() {
+	inodesToDrop := uint64(0)
+
+	vS.Lock()
+
+	if (vS.inodeCacheLRUItems * globals.inodeSize) > vS.inodeCacheLRUMaxBytes {
+		// Check, at most, 1.25 * (minimum_number_to_drop)
+		inodesToDrop = (vS.inodeCacheLRUItems * globals.inodeSize) - vS.inodeCacheLRUMaxBytes
+		inodesToDrop = inodesToDrop / globals.inodeSize
+		inodesToDrop += inodesToDrop / 4
+	}
+	for ((vS.inodeCacheLRUItems * globals.inodeSize) > vS.inodeCacheLRUMaxBytes) &&
+		(inodesToDrop > 0) {
+		inodesToDrop--
+
+		ic := vS.inodeCacheLRUHead
+
+		// Create a DLM lock object
+		id := dlm.GenerateCallerID()
+		inodeRWLock, _ := InitInodeLock(vS.volumeName, ic.InodeNumber, id)
+		err := inodeRWLock.TryWriteLock()
+
+		// Inode is locked; skip it
+		if err != nil {
+			// Move inode to tail of LRU
+			vS.inodeCacheTouchWhileLocked(ic)
+			continue
+		}
+
+		if (ic.inFlightLogSegmentMap != nil) && (ic.dirty == false) {
+			// The inode is busy - drop the DLM lock and move to tail
+			inodeRWLock.Unlock()
+			vS.inodeCacheTouchWhileLocked(ic)
+			continue
+		}
+
+		vS.Unlock()
+
+		var ok bool
+
+		ok, err = vS.inodeCacheDrop(ic)
+		if err != nil || !ok {
+			pStr := fmt.Errorf("The inodes was not found in the inode cache - ok: %v err: %v", ok, err)
+			panic(pStr)
+		}
+
+		// TODO - I am releasing the locks out of order here.   That seems wrong.
+		// We acquire the locks in this order:
+		// 1. Volume lock
+		// 2. DLM lock for inode
+		//
+		// We then release the volume lock before deleting the inode from the cache and
+		// then releasing the DLM lock.  Is this okay?
+		inodeRWLock.Unlock()
+
+		// vS.inodeCacheDrop() removed the inode from the freelist so
+		// the head is now different
+		vS.Lock()
+	}
+	vS.Unlock()
+}
+
 func (vS *volumeStruct) inodeCacheDropWhileLocked(inode *inMemoryInodeStruct) (ok bool, err error) {
 	ok, err = vS.inodeCache.DeleteByKey(inode.InodeNumber)
 	if (nil != err) || !ok {
 		return
 	}
 
-	// TODO: Remove inode from inodeCacheLRU
+	// QUESTION - Is the above TODO still correct?????
```

Comment:
Yeah - I must have forgotten to remove this TODO - since I implemented it.

BTW, personal preference... I would have much preferred you actually "called" the routines I wrote (as you would have had to do had you converted to list.List, right?)... I really don't see why you needed to "embed" my clean/simple func inside your more complex function. At least that's the way it appears in the GitHub/Gerrit-like review tool. Why didn't you just call my function?

Reply:
Must be a github interface problem - I am only calling your functions unless I am missing something